proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -507,39 +507,1153 @@ locale delete_one = delete_one_conc + delete_one_abs +
+     apply (erule sc_at_cross; clarsimp simp add: obj_at_def is_sc_obj; simp?)
+     apply (erule valid_objs_valid_sched_context_size, fastforce)
+    apply (rule conjI)
+     apply (fastforce dest!: st_tcb_at_coerce_concrete simp: thread_state_relation_def op_equal)
+    apply (rule conjI)
+     apply (fastforce dest!: state_refs_of_cross_eq)
+    apply (fastforce dest: state_relationD simp: sc_replies_relation_sc_with_reply_cross_eq)
+   apply (clarsimp simp: corres_underlying_def)
+   apply (rule conjI)
+    defer
+    apply (erule no_failD[OF no_fail_updateReply])
+   apply (clarsimp simp: update_sched_context_def updateReply_def getReply_def
+                         setReply_def getObject_def setObject_def in_monad split_def
+                         get_object_def set_object_def bind_assoc loadObject_default_def
+                         obj_at'_def projectKOs reply_sc_reply_at_def obj_at_def exec_gets
+                         updateObject_default_def in_magnitude_check objBits_simps' a_type_def)
+   apply (clarsimp simp: exec_get put_def)
+   apply (rename_tac reply' sc' nextr)
+   apply (prop_tac "reply_at' nrp b")
+    apply (clarsimp simp: obj_at'_def projectKOs objBits_simps')
+   apply (prop_tac "reply_at nrp a")
+    apply (drule (1) valid_sched_context_objsI)
+    apply (clarsimp simp: valid_sched_context_def)
+    apply (frule_tac nrp=nrp in next_reply_in_sc_replies[where rp=rp, OF state_relation_sc_replies_relation])
+       apply (simp add: obj_at'_def projectKOs objBits_simps' opt_map_left_Some)+
+    apply (clarsimp simp: vs_heap_simps)
+   apply (drule_tac x=nextr in z)
+   apply (clarsimp simp: state_relation_def obj_at_def is_reply obj_at'_def projectKOs)
+   apply (rule conjI)
+    (* pspace_relation *)
+    apply (simp only: pspace_relation_def simp_thms
+                      pspace_dom_update[where x="kernel_object.SchedContext _ _"
+                                          and v="kernel_object.SchedContext _ _",
+                                        simplified a_type_def, simplified])
+    apply (simp only: dom_fun_upd2 simp_thms)
+    apply (elim conjE)
+    apply (frule bspec, erule domI)
+    apply (frule_tac x=nrp in bspec, erule domI)
+    apply (rule ballI, drule(1) bspec)
+    apply (drule domD)
+    apply (prop_tac "scp \<noteq> nrp", clarsimp)
+    apply (clarsimp simp: project_inject split: if_split_asm kernel_object.split_asm)
+    apply (rule conjI; clarsimp)
+     apply (clarsimp simp: sc_relation_def)
+    apply (rename_tac bb aa ba)
+    apply (drule_tac x="(aa, ba)" in bspec, simp)
+    apply (clarsimp simp: obj_at_def is_reply)
+    apply (frule_tac ko'="kernel_object.Reply reply" and x'=nrp in obj_relation_cut_same_type)
+       apply simp+
+    apply (clarsimp simp: reply_relation_def)
+   apply (rule conjI)
+    (* sc_replies_relation *)
+    apply (clarsimp simp: projectKO_opt_sc opt_map_left_Some)
+    apply (frule (2) StateRelation.sc_replies_relation_scReply)
```

Comment:
The qualification here isn't needed (though maybe it's better to have it for some reason I don't know of)